pkg/apiserver/certificate/certificate.go

```diff
@@ -110,5 +107,84 @@ func ApplyServerCert(selfSignedCert bool, client kubernetes.Interface, aggregato
 	}
 
 	caCertController := newCACertController(caContentProvider, client, aggregatorClient)
+
+	if selfSignedCert {
+		go rotateSelfSignedCertificates(caCertController, secureServing, globalMaxRotateDuration)
+
+	}
+
 	return caCertController, nil
 }
+
+// generateSelfSignedCertificate generates a new self signed certificate.
+func generateSelfSignedCertificate(secureServing *options.SecureServingOptionsWithLoopback) (dynamiccertificates.CAContentProvider, error) {
+	var err error
+	var caContentProvider dynamiccertificates.CAContentProvider
+
+	// Set the PairName but leave certificate directory blank to generate in-memory by default.
+	secureServing.ServerCert.CertDirectory = ""
+	secureServing.ServerCert.PairName = "antrea-controller"
+
+	if err := secureServing.MaybeDefaultWithSelfSignedCerts("antrea", GetAntreaServerNames(), []net.IP{net.ParseIP("127.0.0.1")}); err != nil {
+		return nil, fmt.Errorf("error creating self-signed certificates: %v", err)
+	}
+
+	certPEMBlock, _ := secureServing.ServerCert.GeneratedCert.CurrentCertKeyContent()
+	caContentProvider, err = dynamiccertificates.NewStaticCAContent("self-signed cert", certPEMBlock)
+	if err != nil {
+		return nil, fmt.Errorf("error reading self-signed CA certificate: %v", err)
+	}
+
+	return caContentProvider, nil
+}
+
+// Used to determine which is sooner, the provided rotateInterval or the expiration date
+// of the cert. Used to allow for unit testing with a far shorter rotation period.
+// Also can be used to pass a user provided rotation window.
+func nextRotationDuration(secureServing *options.SecureServingOptionsWithLoopback,
+	rotateInterval time.Duration) (time.Duration, error) {
+	pemCert, pemKey := secureServing.ServerCert.GeneratedCert.CurrentCertKeyContent()
+	cert, err := tls.X509KeyPair(pemCert, pemKey)
+	if err != nil {
+		return time.Duration(0), err
+	}
+
+	x509Cert, err := x509.ParseCertificate(cert.Certificate[0])
+	if err != nil {
+		return time.Duration(0), err
```

Comment:
I suggest wrapping the error here just in case. I believe the error message would be quite cryptic otherwise.
```
		return time.Duration(0), fmt.Errorf("error parsing generated certificate: %v", err)
```